Fix library build with OpenSSL 4.x

MySQL is preparing to support OpenSSL 4.x (currently in beta, expected to become GA this month) in one of the future releases. Unfortunately, libfido2 cannot be built with OpenSSL 4.0 due some API changes.

Example build errors (using Linux, OpenSSL 4.0.0 alpha1) below apply to libfido2 v1.15.0, but looking at the latest libfido2 code, same issues are also applicable to latest master as well:

```
/w/mrajcic/mysql/extra/libfido2/libfido2-1.15.0/src/ecdh.c: In function ‘hkdf_sha256’:
/w/mrajcic/mysql/extra/libfido2/libfido2-1.15.0/src/ecdh.c:51:19: error: implicit declaration of function ‘EVP_MD_meth_dup’; did you mean ‘EVP_MD_CTX_dup’? [-Wimplicit-function-declaration]
   51 |             (md = EVP_MD_meth_dup(const_md)) == NULL ||
      |                   ^~~~~~~~~~~~~~~
      |                   EVP_MD_CTX_dup
/w/mrajcic/mysql/extra/libfido2/libfido2-1.15.0/src/ecdh.c:51:17: error: assignment to ‘EVP_MD *’ {aka ‘struct evp_md_st *’} from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
   51 |             (md = EVP_MD_meth_dup(const_md)) == NULL ||
      |                 ^
/w/mrajcic/mysql/extra/libfido2/libfido2-1.15.0/src/ecdh.c:72:17: error: implicit declaration of function ‘EVP_MD_meth_free’; did you mean ‘EVP_MD_CTX_free’? [-Wimplicit-function-declaration]
   72 |                 EVP_MD_meth_free(md);
      |                 ^~~~~~~~~~~~~~~~
      |                 EVP_MD_CTX_free

```

The patch below has been proposed by GPT-5.2, feel free to use it. It fixes the compile issue, but I did not run any libfido2 tests to check for validity.

--- a/extra/libfido2/libfido2-1.15.0/src/ecdh.c
+++ b/extra/libfido2/libfido2-1.15.0/src/ecdh.c
```
@@ -35,8 +35,7 @@ hkdf_sha256(uint8_t *key, const char *info, const fido_blob_t *secret)
 static int
 hkdf_sha256(uint8_t *key, char *info, fido_blob_t *secret)
 {
-	const EVP_MD *const_md;
-	EVP_MD *md = NULL;
+	const EVP_MD *md;
 	EVP_PKEY_CTX *ctx = NULL;
 	size_t keylen = SHA256_DIGEST_LENGTH;
 	uint8_t	salt[32];
@@ -47,8 +46,7 @@ hkdf_sha256(uint8_t *key, char *info, fido_blob_t *secret)
 		fido_log_debug("%s: invalid param", __func__);
 		goto fail;
 	}
-	if ((const_md = EVP_sha256()) == NULL ||
-	    (md = EVP_MD_meth_dup(const_md)) == NULL ||
+	if ((md = EVP_sha256()) == NULL ||
 	    (ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL)) == NULL) {
 		fido_log_debug("%s: init", __func__);
 		goto fail;
@@ -68,8 +66,6 @@ hkdf_sha256(uint8_t *key, char *info, fido_blob_t *secret)
 
 	ok = 0;
 fail:
-	if (md != NULL)
-		EVP_MD_meth_free(md);
 	if (ctx != NULL)
 		EVP_PKEY_CTX_free(ctx);
```
 
I hope this will help you to adapt libfido2 to work with OpenSSL 4.x.

--- a/src/ecdh.c
+++ b/src/ecdh.c
@@ -37,8 +37,7 @@ int
 hkdf_sha256(uint8_t *key, size_t keylen, const char *info,
     const fido_blob_t *secret)
 {
-	const EVP_MD *const_md;
-	EVP_MD *md = NULL;
+	const EVP_MD *md;
 	EVP_PKEY_CTX *ctx = NULL;
 	uint8_t salt[32];
 	int ok = -1;
@@ -48,8 +47,7 @@ hkdf_sha256(uint8_t *key, size_t keylen, const char *info,
 		fido_log_debug("%s: invalid param", __func__);
 		goto fail;
 	}
-	if ((const_md = EVP_sha256()) == NULL ||
-	    (md = EVP_MD_meth_dup(const_md)) == NULL ||
+	if ((md = EVP_sha256()) == NULL ||
 	    (ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, NULL)) == NULL) {
 		fido_log_debug("%s: init", __func__);
 		goto fail;
@@ -70,8 +68,6 @@ hkdf_sha256(uint8_t *key, size_t keylen, const char *info,
 
 	ok = 0;
 fail:
-	if (md != NULL)
-		EVP_MD_meth_free(md);
 	if (ctx != NULL)
 		EVP_PKEY_CTX_free(ctx);
 

